### PR TITLE
refactor(core): Refactor the stream.go closing path

### PR DESCRIPTION
**Pull request by timoffex**
Description
---
Removes the confusingly named `outChan` field which was only used in `FinishAndClose`, likely to ensure that the exit record was fully processed before the stream closed its `loopBackChan`. Replaces this with a `done` channel that's closed instead of closing `inChan` / `loopBackChan`.

### CODE REVIEW

**Comment by dmitryduev** [comment]
the main purpose of the ctx was to pass it into sender to be able to cancel any in-flight requests that are waiting to be retried (ran out into this issue at some point) when it's cancelled in the stream. i thought there was a comment about it...

**Comment by timoffex** [comment]
> the main purpose of the ctx was to pass it into sender to be able to cancel any in-flight requests that are waiting to be retried (ran out into this issue at some point) when it's cancelled in the stream. i thought there was a comment about it...

it only cancels at the end of the defer state machine or in some mailbox logic. Stream was relying on the context cancellation to determine whether the Handler-Writer-Sender loop is complete or if messages are still being processed.

**Comment by timoffex** [comment]
@dmitryduev - simplified this PR a bit

File: core/pkg/server/stream.go [hunk 1]
```diff
@@ -342,83 +330,94 @@ func (s *Stream) Start() {
...
+
+		// Otherwise, race between inChan / loopBackChan / done.
+		default:
+			select {
+			case x := <-s.inChan:
+				handlerChan <- x
+			case x := <-s.loopBackChan:
+				handlerChan <- x
+			case <-s.done:
+				return
```
Comment (kptkin) [inline core/pkg/server/stream.go:366]: i think ideally we want to guarantee that we processed all the data before closing the stream?

File: core/pkg/server/stream.go [hunk 2]
```diff
@@ -342,83 +330,94 @@ func (s *Stream) Start() {
 	}()
 
 	// handle dispatching between components
-	s.wg.Add(1)
+	s.wg.Add(2)
```
Comment (kptkin) [inline core/pkg/server/stream.go:333]: don't you need 3 wg?

File: core/pkg/server/stream.go [hunk 3]
```diff
@@ -342,83 +330,94 @@ func (s *Stream) Start() {
...
 	go func() {
```
Comment (kptkin) [inline core/pkg/server/stream.go:334]: do we even need this outer go?

**Review by kptkin — APPROVED** [review]

**Review by dmitryduev — APPROVED** [review]
beautiful!

File: core/pkg/server/stream.go (hunk 3 above)
Comment (timoffex) [reply core/pkg/server/stream.go]: nope, removed!

File: core/pkg/server/stream.go (hunk 1 above)
Comment (timoffex) [reply core/pkg/server/stream.go:364]: the outer select consumes all buffered data before considering the `done` channel. We only close `done` after Sender cancels the `ctx`, so only after the defer state machine has completed---this means there should not be any more requests.

**Pull request merged**